proposals/0000-midstream-bindists.md

```diff
@@ -0,0 +1,199 @@
+* https://github.com/commercialhaskell/stack/issues/6142
+
+Recently, cabal-install had issues with i386 binaries and alpine, delaying a GHCup metadata PR:
+
+* https://github.com/haskell/ghcup-metadata/pull/127#issuecomment-1766020410
+
+These issues are frequent and so far the GHCup developers used to single handedly fix all those missing bindists manually
+and provide them here: https://downloads.haskell.org/~ghcup/unofficial-bindists/
+
+This is unfunded and significant work.
+
+### Bindist maintenance
+
+Sometimes, bindists are broken, e.g. for GHC there are a couple of instances:
+
+* 9.0.2 shipping without profiling info: https://gitlab.haskell.org/ghc/ghc/-/issues/21841
+* DESTDIR variable ignored by `make install`: https://gitlab.haskell.org/ghc/ghc/-/issues/19646
+
+Sometimes, bindists have been built for very old version of linux distros and won't run well on newer linux versions.
+This is currently a problem since Debian has removed ncurses5:
+
+* https://github.com/haskell/ghcup-hs/issues/902
+* https://answers.launchpad.net/ubuntu/+source/ncurses/+question/707838
+* https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1025964
+
+For cabal, there has been the infamous hLock issue:
+
+* https://github.com/haskell/cabal/issues/7313
+* https://github.com/haskell/cabal/issues/7950
+
+This shows that bindists, for current and historical versions, need continuous maintenance. However, upstream developers
+so far have very rarely engaged in this type of maintenance work, pushing it down to GHCup. As an example, here are all the
+manually patched and re-packaged bindists that fix the DESTDIR bug outlined above: https://downloads.haskell.org/ghcup/unofficial-bindists/ghc/curated/
+
+This type of work also requires significant time.
+
+### Quality gateway
+
+There's something pleasing about upstream providing bindists: there's a perceived trust about it in the community,
+e.g. the [haskell.org committee expressed concerns about GHCup changing bindists in the past](https://github.com/haskell-infra/www.haskell.org/issues/212#issuecomment-1272312911).
+
+However, rarely do upstream developers have signifcant experience in redistribution, nor do they have the time to focus
+on all the issues that come with it. Bindists are mostly provided "as-is" and support beyond what the release CI outputs is
+left to midstream (GHCup, stack, ...).
+
+Conceptually, it is a good idea to separate concerns: upstream provides the sources and has tested it. Distributions build the binaries,
+validate that the program can be successfully built from source and make sure that the final artifacts pass the test suite.
+This is good, because we want to know whether end users can build e.g. a functioning GHC from source. If only the release CI outputs
+a GHC that passes the test suite, then something is fundamentally broken.
+
+Distributors are often closer to the end-users and can provide additional support and efforts for the installation experience.
```

Comment:
... but again, I do think this should go upstream as much as possible!

Reply:
It can. One of the goals of this proposal is to feed back such information (e.g. test reports) back to upstream, but in a more structured manner, while still maintaining strict separation of the projects.